pkg/proxy/handlers/spacelister.go

```diff
@@ -36,25 +37,92 @@ func NewSpaceLister(app application.Application) *SpaceLister {
 }
 
 func (s *SpaceLister) HandleSpaceListRequest(ctx echo.Context) error {
+	userID, _ := ctx.Get(context.SubKey).(string)
+	username, _ := ctx.Get(context.UsernameKey).(string)
+
+	userSignup, err := s.GetSignupFunc(nil, userID, username, false)
+	if err != nil {
+		ctx.Logger().Error(errs.Wrap(err, "error retrieving userSignup"))
+		return errorResponse(ctx, apierrors.NewInternalError(err))
+	}
+	if userSignup == nil || userSignup.CompliantUsername == "" {
+		// account exists but the compliant username is not set yet, meaning it has not been fully provisioned yet
+		// return empty workspace list
+		return listWorkspaceResponse(ctx, []toolchainv1alpha1.Workspace{})
+	}
 
+	// list all user workspaces
 	workspaces, err := s.ListUserWorkspaces(ctx)
 	if err != nil {
 		return errorResponse(ctx, apierrors.NewInternalError(err))
 	}
+	return listWorkspaceResponse(ctx, workspaces)
+}
 
+func (s *SpaceLister) HandleSpaceGetRequest(ctx echo.Context) error {
+	userID, _ := ctx.Get(context.SubKey).(string)
+	username, _ := ctx.Get(context.UsernameKey).(string)
 	workspaceName := ctx.Param("workspace")
-	doGetWorkspace := len(workspaceName) > 0
 
-	if doGetWorkspace { // specific workspace requested
-		if len(workspaces) != 1 {
-			// not found
-			r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
-			return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
-		}
-		return getWorkspaceResponse(ctx, workspaces[0])
+	userSignup, err := s.GetSignupFunc(nil, userID, username, false)
+	if err != nil {
+		ctx.Logger().Error(errs.Wrap(err, "error retrieving userSignup"))
+		return errorResponse(ctx, apierrors.NewInternalError(err))
 	}
+	if userSignup == nil || userSignup.CompliantUsername == "" {
+		// account exists but the compliant username is not set yet, meaning it has not been fully provisioned yet
+		// return not found response when specific workspace request was issued
+		r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+		return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
 
-	return listWorkspaceResponse(ctx, workspaces)
+	}
+
+	// get specific workspace
+	workspace, err := s.GetUserWorkspace(ctx, userSignup)
```

Comment:
Both handlers `HandleSpaceListRequest` and the newly added `HandleSpaceGetRequest` need to retrieve the usersignup in order to get the `signup.CompliantUsername` to be used as MUR name for listing the spacebindings. The _get_ and the _list_ are now in different handlers so both of them need to retrieve the usersignup, but maybe I'm missing something.

Reply:
I'm idiot - I selected a wrong code and copy-pasted a wrong method name when doing the comment. Sorry about that. I got lost in the changes between IDE and GH.
I wanted to comment on 
`HandleSpaceListRequest` part:
https://github.com/codeready-toolchain/registration-service/blob/1305bcf1a4583893eb6a7415ba998abc95fad2d6/pkg/proxy/handlers/spacelister.go#L39-L55

you check the UserSignup there and then call the `ListUserWorkspaces` method which does the same again:
https://github.com/codeready-toolchain/registration-service/blob/1305bcf1a4583893eb6a7415ba998abc95fad2d6/pkg/proxy/handlers/spacelister.go#L128-L140

Reply:
Oh no worries, I see now the issue 🤦‍♂️ . I haven't noticed I duplicated this code while refactoring , thanks for pointing that out! Let me fix it!

Reply:
I've moved the usersignup retrieveal inside the `ListUserWorkspaces` and `GetUserWorkspace` functions, so that they are the same. I removed that from the handler functions, mainly because the `ListUserWorkspaces` function is used also by the proxy which doesn't have the usersignup available to be passed as parameter.

Plz see: https://github.com/codeready-toolchain/registration-service/pull/348/commits/01759250a7329b5df84baeca99ea984bd28029ec

Reply:
And I've also added sub tests in HandleSpaceGetRequest for the "get signup error" and "signup has no compliant username set"